pkg/core/dao/dao.go

```diff
@@ -686,46 +688,52 @@ func (dao *Simple) StoreHeaderHashes(hashes []util.Uint256, height uint32) error
 // or in the list of conflicting transactions. If non-zero signers are specified,
 // then additional check against the conflicting transaction signers intersection
 // is held. Do not omit signers in case if it's important to check the validity
-// of a supposedly conflicting on-chain transaction.
-func (dao *Simple) HasTransaction(hash util.Uint256, signers []transaction.Signer) error {
+// of a supposedly conflicting on-chain transaction. The retrieved conflict isn't
+// checked against the maxTraceableBlocks setting if signers are omitted.
+// HasTransaction does not consider the case of block executable.
+func (dao *Simple) HasTransaction(hash util.Uint256, signers []transaction.Signer, currentIndex uint32, maxTraceableBlocks uint32) error {
 	key := dao.makeExecutableKey(hash)
 	bytes, err := dao.Store.Get(key)
 	if err != nil {
 		return nil
 	}
 
-	if len(bytes) < 6 {
+	if len(bytes) < 5 { // (storage.ExecTransaction + index) for conflict record
 		return nil
 	}
-	if bytes[5] != transaction.DummyVersion {
-		return ErrAlreadyExists
+	if len(bytes) != 5 {
+		return ErrAlreadyExists // fully-qualified transaction
 	}
 	if len(signers) == 0 {
 		return ErrHasConflicts
 	}
 
-	sMap := make(map[util.Uint160]struct{}, len(signers))
+	index := binary.LittleEndian.Uint32(bytes[1:])
+	if !isTraceableBlock(index, currentIndex, maxTraceableBlocks) {
+		// The most fresh conflict record is already outdated.
+		return nil
+	}
+
+	sKey := make([]byte, len(key)+util.Uint160Size)
+	copy(sKey, key)
 	for _, s := range signers {
-		sMap[s.Account] = struct{}{}
-	}
-	br := io.NewBinReaderFromBuf(bytes[6:])
-	for {
-		var u util.Uint160
-		u.DecodeBinary(br)
-		if br.Err != nil {
-			if errors.Is(br.Err, iocore.EOF) {
-				break
+		copy(sKey[len(key):], s.Account.BytesBE())
+		v, err := dao.Store.Get(sKey)
+		if err == nil {
+			index = binary.LittleEndian.Uint32(v[1:])
+			if isTraceableBlock(index, currentIndex, maxTraceableBlocks) {
+				return ErrHasConflicts
 			}
-			return fmt.Errorf("failed to decode conflict record: %w", err)
-		}
-		if _, ok := sMap[u]; ok {
-			return ErrHasConflicts
 		}
 	}
 
 	return nil
 }
 
+func isTraceableBlock(index, height, maxTraceableBlocks uint32) bool {
```

Comment:
The way you use it `index` can be a `[]byte`.

Reply:
But we still have to convert it to the proper integer to compare with allowed bounds.

Reply:
Fixed.